From e0fcba7b8f32db6ff116cd90f21fc7ccd413e730 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 11 May 2023 15:39:54 +0300
Subject: [PATCH] [HACK] thermal: k3_j72xx_bandgap: fix probe on am64 platform

AM64 does not allow access to the physical address that on J721 maps to
WKUP_SPARE_FUSE0. Therefore ignore errors during iomap when i2128 does
not apply.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/thermal/k3_j72xx_bandgap.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/thermal/k3_j72xx_bandgap.c b/drivers/thermal/k3_j72xx_bandgap.c
index 6d1d4e965ea1..d4053e0524bf 100644
--- a/drivers/thermal/k3_j72xx_bandgap.c
+++ b/drivers/thermal/k3_j72xx_bandgap.c
@@ -507,13 +507,20 @@ static int k3_j72xx_bandgap_probe(struct platform_device *pdev)
 
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 2);
 	bgp->fuse_base = devm_ioremap_resource(dev, res);
-	if (IS_ERR(bgp->fuse_base))
-		return PTR_ERR(bgp->fuse_base);
 
 	driver_data = of_device_get_match_data(dev);
 	if (driver_data)
 		workaround_needed = driver_data->has_errata_i2128;
 
+	if (IS_ERR(bgp->fuse_base) && !workaround_needed) {
+		dev_warn(dev, "failed mapping J721 WKUP_SPARE_FUSE0 (memory resource #2): %i\n", PTR_ERR(bgp->fuse_base));
+		dev_warn(dev, "this platform doesn't have errata i2128, ignoring error!\n");
+		bgp->fuse_base = NULL;
+	} else if (IS_ERR(bgp->fuse_base)) {
+		dev_err(dev, "failed mapping J721 WKUP_SPARE_FUSE0 (memory resource #2): %i\n", PTR_ERR(bgp->fuse_base));
+		return PTR_ERR(bgp->fuse_base);
+	}
+
 	pm_runtime_enable(dev);
 	ret = pm_runtime_get_sync(dev);
 	if (ret < 0) {
-- 
2.35.3

